upstairs/src/lib.rs

```diff
@@ -5951,34 +5951,8 @@ impl Upstairs {
             return Err(());
         }
 
-        /*
-         * Get the next ID for the guest work struct we will make at the
-         * end. This ID is also put into the IO struct we create that
-         * handles the operation(s) on the storage side.
-         */
-        let mut gw = self.guest.guest_work.lock().await;
-        let mut downstairs = self.downstairs.lock().await;
         let ddef = self.ddef.lock().await.get_def().unwrap();
```

Comment:
The lock ordering here might cause us a problem.
All the other places where we get the three locks (gw, downstairs, ddef), they always go in
that order.  If we get ddef here, but someone else has gw and or ds and then tries to get
ddef, we will deadlock.

Some options,
1) If we need ddef, maybe we move it up to the first lock selected?
2) For this use case, we might be able to get the fields we need out of ddef and
then release the lock?  At the moment the `ddef` fileds we use here should not be
changing once they are initially set.  At some point we may want to grow a region, and
we will have to deal with that, but with the current setup, we set it once and panic if it
ever changes.

Reply:
The `ddef` lock isn't actually held here: `get_def()` gets a copy of the `RegionDefinition`, and then the lock is dropped at the end of the line (so we're already doing your Option 2).

Reply:
Ah, well then, no deadlock.

We might have to give this area some attention if we change how the `ddef` is updated, but we can
burn that bridge when we get to it.